src/main/rx/rx.c

```diff
@@ -377,23 +377,55 @@ static uint16_t calculateNonDataDrivenChannel(uint8_t chan, uint16_t sample)
     return rcDataMean[chan] / PPM_AND_PWM_SAMPLE_COUNT;
 }
 
-static uint16_t getRxfailValue(uint8_t channel)
+int16_t rxValueStepToTarget(int16_t currentRxValue, uint16_t targetRxValue) {
+
+#define RX_FAIL_STEP_SIZE 30
+#define RX_FAIL_STEP_DEADBAND 50
+
+    if ((currentRxValue >= (targetRxValue - RX_FAIL_STEP_DEADBAND)) && (currentRxValue < targetRxValue)) {
+        return targetRxValue;
+    } else if ((currentRxValue <= (targetRxValue + RX_FAIL_STEP_DEADBAND)) && (currentRxValue > targetRxValue)) {
+        return targetRxValue;
+    } else if (currentRxValue < (targetRxValue - RX_FAIL_STEP_DEADBAND)) {
+        return (currentRxValue + RX_FAIL_STEP_SIZE);
+    } else if (currentRxValue > (targetRxValue + RX_FAIL_STEP_DEADBAND)) {
+        return (currentRxValue - RX_FAIL_STEP_SIZE);
+    } else {
+        return targetRxValue;
+    }
+}
+
+static int16_t getRxfailValue(uint8_t channel)
 {
     rxFailsafeChannelConfiguration_t *channelFailsafeConfiguration = &rxConfig->failsafe_channel_configurations[channel];
 
+#define RX_FAIL_STEP_TIME_BEFORE_STEPPING 1500
+
     switch(channelFailsafeConfiguration->mode) {
         case RX_FAILSAFE_MODE_AUTO:
             switch (channel) {
                 case ROLL:
                 case PITCH:
                 case YAW:
-                    return rxConfig->midrc;
+                	if (millis() < RX_FAIL_STEP_TIME_BEFORE_STEPPING) {
+                        return rxConfig->midrc;
+                    } else {
+                        return rxValueStepToTarget(rcData[channel], rxConfig->midrc);
```

Comment:
Thats was my impression too. Took a while to dig into it to fully
understand it.

Too many special cases.

The true is that we do have many reported serial glitches and I think some
of them might lead into disarming for users who use arm/disarm on the
switch as the current rxfail mechanism acts to abrupt.
Just guessing here based on some cases.
Op 14 okt. 2015 17:44 schreef "Petr Ledvina" <notifications@github.com>:

> In src/main/rx/rx.c
> <https://github.com/borisbstyle/cleanflight/pull/19#discussion_r42011259>:
>
> >      switch(channelFailsafeConfiguration->mode) {
> >          case RX_FAILSAFE_MODE_AUTO:
> >              switch (channel) {
> >                  case ROLL:
> >                  case PITCH:
> >                  case YAW:
> > -                    return rxConfig->midrc;
> > +                	if (millis() < RX_FAIL_STEP_TIME_BEFORE_STEPPING) {
> > +                        return rxConfig->midrc;
> > +                    } else {
> > +                        return rxValueStepToTarget(rcData[channel], rxConfig->midrc);
>
> Whole RX-handling logic is getting far too complicated. Adding new special
> cases will make it even worse ...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/borisbstyle/cleanflight/pull/19/files#r42011259>.
>